tpcx_bb/queries/q27/tpcx_bb_query_27_sql.py

```diff
@@ -0,0 +1,123 @@
+# limitations under the License.
+#
+
+import sys
+
+from xbb_tools.text import (
+    create_sentences_from_reviews,
+    create_words_from_sentences
+)
+
+from xbb_tools.cluster_startup import attach_to_cluster
+from dask.distributed import wait
+import spacy
+
+from xbb_tools.utils import (
+    benchmark,
+    tpcxbb_argparser,
+    run_query,
+)
+
+
+# -------- Q27 -----------
+q27_pr_item_sk = 10002
+EOL_CHAR = "."
+
+
+def read_tables(data_dir, bc):
+    bc.create_table("product_reviews", data_dir + "/product_reviews/*.parquet")
+
+
+def ner_parser(df, col_string, batch_size=256):
+    spacy.require_gpu()
+    nlp = spacy.load("en_core_web_sm")
+    docs = nlp.pipe(df[col_string], disable=["tagger", "parser"], batch_size=batch_size)
+    out = []
+    for doc in docs:
+        l = [ent.text for ent in doc.ents if ent.label_ == "ORG"]
+        val = ", "
+        l = val.join(l)
+        out.append(l)
+    df["company_name_list"] = out
+    return df
+
+
+def main(data_dir, client, bc, config):
+    benchmark(read_tables, data_dir, bc, dask_profile=config["dask_profile"])
+
+    import dask_cudf
+
+    query = f"""
+        SELECT pr_review_sk, pr_item_sk, pr_review_content
+        FROM product_reviews
+        WHERE pr_item_sk = {q27_pr_item_sk}
+    """
+    product_reviews_df = bc.sql(query)
+
+    sentences = product_reviews_df.map_partitions(
+        create_sentences_from_reviews,
+        review_column="pr_review_content",
+        end_of_line_char=EOL_CHAR,
+    )
+
+    # need the global position in the sentence tokenized df
+    sentences["x"] = 1
+    sentences["sentence_tokenized_global_pos"] = sentences.x.cumsum()
+    del sentences["x"]
+
+    sentences = sentences.persist()
+    wait(sentences)
+
+    # Do the NER
+    sentences = sentences.to_dask_dataframe()
+    ner_parsed = sentences.map_partitions(ner_parser, "sentence")
+    ner_parsed = dask_cudf.from_dask_dataframe(ner_parsed)
+    ner_parsed = ner_parsed.persist()
+    wait(ner_parsed)
+
+    ner_parsed = ner_parsed[ner_parsed.company_name_list != ""]
+
+    # separate NER results into one row per found company
+    repeated_names = ner_parsed.map_partitions(
+        create_words_from_sentences,
+        sentence_column="company_name_list",
+        global_position_column="sentence_tokenized_global_pos",
+        delimiter="é",
+    )
+
+    # recombine
+    bc.create_table('repeated_names', repeated_names)
+    bc.create_table('ner_parsed', ner_parsed)
+
+    query = f"""
+        SELECT review_idx_global_pos as review_sk,
+            CAST({q27_pr_item_sk} AS BIGINT) as item_sk,
+            word as company_name,
+            sentence as review_sentence
+        FROM repeated_names left join ner_parsed
+        ON sentence_idx_global_pos = sentence_tokenized_global_pos
+        ORDER BY review_idx_global_pos, item_sk, word, sentence
+    """
+    recombined = bc.sql(query)
+    return recombined.compute()
```

Comment:
We should probably not call `compute` here. This will bring everything to the client process. Can we persist and write a distributed output?

Reply:
Thanks for reviewing @beckernick, I just removed the compute call. Would it be necessary as you mention to add the call to persist? since there is no ML processing afterward? or in general, is it for good practice?

Reply:
Great question. It's not strictly necessary, but we should still do it because logically separating the computation for the "business logic" section from the "writing data" section is useful for performance analysis.

Since persist is non-blocking, this would also need to use `wait` to actually accomplish the above. Something like the following would probably work:

```
recombined = bc.sql(query)
recombined = recombined.persist()
wait(recombined)
return recombined
```

If `bc.sql(query)` already triggers computation (I believe it doesn't), then you could skip the persist and just use wait.

Reply:
Got it, thanks for the clarification.